File: parent/3517 -- Smallest Palindromic Rearrangement I/mine3.go
package main

import (
	"strings"
)

func smallestPalindrome(s string) string {
	var freq [26]int
	var res []string
	var mid string

	for _, char := range s {
		freq[int(char)-int('a')] += 1
	}

	for i, count := range freq {
		char := string(i + int('a'))
		if count > 0 {
			res = append(res, strings.Repeat(char, count/2))
			if count%2 == 1 {
				mid = char
			}
		}
	}

	res = append(res, mid)

	for i := len(freq) - 1; i >= 0; i-- {
		count := freq[i]
		char := string(i + int('a'))

		if count > 0 {
			res = append(res, strings.Repeat(char, count/2))
		}
	}

	return strings.Join(res, "")
}

File: parent/3517 -- Smallest Palindromic Rearrangement I/mine3.py
class Solution:
    def smallestPalindrome(self, s: str) -> str:
        # Counting Sort. Most efficient
        # Time: O(n)
        # Space: O(1)

        freq = [0] * 26
        res = []
        mid = ''

        for char in s:
            freq[ord(char)-ord('a')] += 1
        
        for i, count in enumerate(freq):
            char = chr(i + ord('a'))
            if count > 0: 
                res.append(char * (count//2))
                if count%2 == 1:
                    mid = char
        
        return ''.join(res + [mid] + res[::-1])